From 8a3544f8959375ee167a46ff9e7fc4050127b349 Mon Sep 17 00:00:00 2001
From: Artur Paszkiewicz <artur.paszkiewicz@intel.com>
Date: Mon, 2 Jun 2014 15:02:59 +0200
Subject: [PATCH] imsm: retry load_and_parse_mpb if we suspect mdmon has made
 modifications

If the checksum verification fails in mdadm and mdmon is running, retry
the load to get a consistent snapshot of the mpb.

Based on db575f3b

Signed-off-by: Artur Paszkiewicz <artur.paszkiewicz@intel.com>
Signed-off-by: NeilBrown <neilb@suse.de>
---
 super-intel.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/super-intel.c b/super-intel.c
index f0a7ab5..9dd807a 100644
--- a/super-intel.c
+++ b/super-intel.c
@@ -4422,6 +4422,7 @@ static int load_super_imsm(struct supertype *st, int fd, char *devname)
 {
 	struct intel_super *super;
 	int rv;
+	int retry;
 
 	if (!st->ignore_hw_compat && test_partition(fd))
 		/* IMSM not allowed on partitions */
@@ -4444,6 +4445,22 @@ static int load_super_imsm(struct supertype *st, int fd, char *devname)
 	}
 	rv = load_and_parse_mpb(fd, super, devname, 0);
 
+	/* retry the load if we might have raced against mdmon */
+	if (rv == 3) {
+		struct mdstat_ent *mdstat = mdstat_by_component(fd2devnm(fd));
+
+		if (mdstat && mdmon_running(mdstat->devnm) && getpid() != mdmon_pid(mdstat->devnm)) {
+			for (retry = 0; retry < 3; retry++) {
+				usleep(3000);
+				rv = load_and_parse_mpb(fd, super, devname, 0);
+				if (rv != 3)
+					break;
+			}
+		}
+
+		free_mdstat(mdstat);
+	}
+
 	if (rv) {
 		if (devname)
 			pr_err("Failed to load all information "
-- 
1.9.3

